pkg/preflight/aws.go

```diff
@@ -0,0 +1,162 @@
+	if err != nil {
+		// Volume creation doesn't take volume name, so it's guaranteed there's no conflict
+		logrus.WithError(err).Errorf("failed to create eks volume")
+		return err
+	}
+
+	logrus.WithFields(logrus.Fields{
+		"volume": v,
+	}).Infof("JLIAO: volume created") // TODO: remove
+
+	// Track the created volume in a list, so they can be cleaned up later
+	vol := v.(*ec2.Volume)
+	a.volList[*vol.VolumeId] = true
+
+	// Attach volume
+	_, err = a.ops.Attach(*vol.VolumeId, nil)
+	if err != nil {
+		// Operator container doesn't have access to /dev/ directory on the host, so we expect an error here
+		// unable to map volume vol-03227127b9b7b442e with block device mapping /dev/xvdp to an actual device path on the host
+		errGetPath := fmt.Sprintf("unable to map volume %s with block device mapping", *vol.VolumeId)
+		errAlreadyAttached := fmt.Sprintf("%s is already attached to an instance", *vol.VolumeId)
+		if !strings.Contains(err.Error(), errGetPath) && !strings.Contains(err.Error(), errAlreadyAttached) {
+			logrus.WithError(err).Errorf("failed to attach volume")
+			return err
+		}
+
+		// Try to get the volume attachment info
+		volumes, err := a.ops.Inspect([]*string{vol.VolumeId})
+		if err != nil || len(volumes) != 1 {
+			logrus.WithError(err).Errorf("failed to inspect volume")
+			return err
+		}
+
+		// Inspect volume attachments
+		vol := volumes[0].(*ec2.Volume)
+		logrus.WithFields(logrus.Fields{
+			"vol": vol,
+		}).Infof("JLIAO: vol inspected") // TODO: remove
+		if len(vol.Attachments) != 1 || *vol.Attachments[0].State != ec2.VolumeAttachmentStateAttached {
+			logrus.WithFields(logrus.Fields{
+				"attachments": vol.Attachments,
+			}).Errorf("volume is not attached")
+			return fmt.Errorf("volume is not attached")
+		}
+		logrus.WithFields(logrus.Fields{
+			"volume": vol,
+		}).Infof("JLIAO: volume attached") // TODO: remove
+	}
+
+	// Detach
+	if err := a.ops.Detach(*vol.VolumeId); err != nil {
+		//  IncorrectState: Volume 'vol-0dd22e588c5dfee1e' is in the 'available' state.\n\tstatus code: 400, request id: e8a7f095-138f-4271-840b-c7ae08a89529
+		errAlreadyDetached := fmt.Sprintf("Volume '%s' is in the 'available' state", *vol.VolumeId)
+		if !strings.Contains(err.Error(), errAlreadyDetached) {
+			logrus.WithError(err).Errorf("failed to detach volume")
+			return err
+		}
+	}
+	logrus.Infof("JLIAO: volume detached") // TODO: remove
+
+	if err := a.ops.Delete(*vol.VolumeId); err != nil {
+		logrus.WithError(err).Errorf("failed to delete volume")
+		return err
+	}
+
+	logrus.Infof("JLIAO: volume deleted") // TODO: remove
+	delete(a.volList, *vol.VolumeId)
+
+	// Do a cleanup when Gpreflight passed, as it's guaranteed to have delete permission
+	a.cleanupVolumes()
+	return nil
+}
+
+func (a *aws) cleanupVolumes() {
+	if len(a.volList) == 0 {
+		return
+	}
+
+	logrus.Infof("cleaning up volumes created in preflight check")
+	for vol := range a.volList {
+		err := a.ops.Detach(vol)
+		if err == nil {
```

Comment:
the preflight already passed at this step, so guaranteed to have permission to do those operatorion, we try our best to delete old volumes. for now if it fails we don't do anything.

Reply:
I agree, I'm also concerned about this if the Detach fails for some other reason like temporary AWS API errors, some other weird thing (I dunno if there's a "detach lock" feature), or who knows what, because then we could leave a disk just hanging forever.

Do we have a way to check for preflight disks after the fact to continually keep them cleaned up?

Reply:
idea solution should be when creating the volume, we apply a tag saying it's a preflight check volume, associated with the cluster uid or something. during the cleanup, operator should list volume based on the tag then delete all of them. don't know how complex it is, but definitely a more proper way. That's something we need to do before GA, but can be done in following PRs.

Reply:
Create a ticket to track detach delete failures.